---
title: "Organism.dplyr"
output: 
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Organism.dplyr}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

The `r Biocpkg("Organism.dplyr")` creates an on disk sqlite database to hold 
data of an organism combined from an 'org' package (e.g., org.Hs.eg.db) and a
genome coordinate functionality of the 'TxDb' package (e.g., 
`r Biocpkg("TxDb.Hsapiens.UCSC.hg38.knownGene")`). It aims to provide an 
integrated presentation of identifiers and genomic coordinates. And a 
_src_organism_ object is created to point to the database. 

The _src_organism_ object is created as an extension of _src_sql_ and 
_src_sqlite_ from [dplyr][], which inherited all [dplyr][] methods. It also 
implements the `select()` interface from `r Biocpkg("AnnotationDbi")` and 
genomic coordinates extractors from `r Biocpkg("GenomicFeatures")`.

# Constructing a _src_organism_

## Make sqlite datebase from 'TxDb' package

The `src_organism()` constructor creates an on disk sqlite database file with
data from a given 'TxDb' package and corresponding 'org' package. When dbpath
is given, file is created at the given path, otherwise temporary file is
created.

```{r, echo=FALSE}
suppressPackageStartupMessages({
    library(Organism.dplyr)
    library(ggplot2)
})
```

Running `src_organism()` without a given path will save the sqlite file to a
tempdir():

```{r}
src <- src_organism("TxDb.Hsapiens.UCSC.hg38.knownGene")
```

Alternatively you can provide explicit path to where the sqlite file should
be saved.

```{r, eval=FALSE}
src <- src_organism("TxDb.Hsapiens.UCSC.hg38.knownGene", "path/to/save/sqlite")
```

`supportedOrganisms()` provides a list of organisms with corresponding 'org' 
and 'TxDb' packages being supported.

```{r}
supportedOrganisms()
```

## Make sqlite datebase from organism name

Organism name, genome and id could be specified to create sqlite database. 
Organism name (either Organism or common name) must be provided to create the 
database, if genome and/or id are not provided, most recent 'TxDb' package is 
used.

```{r}
src <- src_ucsc("human")
```

## Access existing sqlite file 

Existing on disk sqlite file can be accessed without recreating the database. 

```{r}
path <- system.file("extdata", package = "Organism.dplyr")
src1 <- src_organism(dbpath = paste0(path, "/example.sqlite"))
src1
```

# Common operations

## Basic operations

All methods from package [dplyr][] can be used for a _src_organism_ object. 

Look at all available tables. 
```{r}
src_tbls(src)
```

Look at data from one specific table. 
```{r}
tbl(src, "id")
```

Look at fields of one table. 
```{r}
colnames(tbl(src, "id"))
```

Below are some examples of querying tables using dplyr. 

1. Gene symbol start with "BRCA"

```{r}
tbl(src, "id") %>% 
    filter(symbol %like% "BRCA%") %>% 
    dplyr::select(entrez, map, ensembl, symbol) %>% 
    distinct() %>% arrange(symbol) %>% collect()
```

2. Go info for gene symbol "PTEN"

```{r}
inner_join(tbl(src, "id"), tbl(src, "id_go")) %>% 
    filter(symbol == "PTEN") %>% 
    dplyr::select(entrez, ensembl, symbol, go, evidence, ontology)
```

3. Genes with more than 100 transcripts

```{r}
txcount <- inner_join(tbl(src, "id"), tbl(src, "ranges_tx")) %>% 
    dplyr::select(symbol, tx_id) %>% 
    group_by(symbol) %>% 
    summarise(count = count(symbol)) %>% 
    filter(count > 100) %>% 
    dplyr::select(symbol, count) %>% 
    arrange(desc(count)) %>% 
    collect(n=Inf)

txcount
```

```{r, echo=FALSE}
ggplot(txcount, aes(x = txcount$symbol[order(txcount$symbol)], 
                    y = txcount$count)) +
    geom_point() + 
    ggtitle("Transcript count") + 
    labs(x = "symbol") + 
    labs(y = "count")
```

4. Gene coordinates of symbol "PTEN" and "BRCA1" as _GRanges_

```{r}
inner_join(tbl(src, "id"), tbl(src, "ranges_gene")) %>% 
    filter(symbol %in% c("PTEN", "BRCA1")) %>% 
    dplyr::select(gene_chrom, gene_start, gene_end, gene_strand, 
                  symbol, map) %>% 
    collect() %>% GenomicRanges::GRanges() 
```

## "select" interface 
Methods `select()`, `keytypes()`, `keys()`, `columns()` and `mapIds` from 
`r Biocpkg("AnnotationDbi")` are implemented for _src_organism_ objects.

### `keytypes()`

Use `keytypes()` to discover which keytypes can be passed to keytype argument 
of methods `select()` or `keys()`.

```{r}
keytypes(src)
```

### `columns()`

Use `columns()` to discover which kinds of data can be returned for the 
_src_organism_ object.

```{r}
columns(src)
```

### `keys()`

`keys()` returns keys for the _src_organism_ object. By default it returns the 
primary keys for the database, and returns the keys from that keytype when the 
keytype argument is used.

Keys of entrez

```{r}
head(keys(src))
```

Keys of symbol

```{r}
head(keys(src, "symbol"))
```

### `select()`

`select()` retrieves the data as a _tibble_ based on parameters for selected 
keys columns and keytype arguments. If requested columns that have multiple 
matches for the keys, `select()` will return a _tibble_ with one row for each 
possible match. 

```{r}
keytype <- "symbol"
keys <- c("PTEN", "BRCA1")
columns <- c("entrez", "tx_id", "tx_name","exon_id")
select(src, keys, columns, keytype)
```

### `mapIds()`

`mapIds()` gets the mapped ids (column) for a set of keys that are of a 
particular keytype. Usually returned as a named character vector.

```{r}
mapIds(src, keys, column = "tx_name", keytype)
```

## Genomic coordinates extractors

Eleven genomic coordinates extractor methods are available in this package: 
`transcripts()`, `exons()`, `cds()`, `genes()`, `promoters()`, 
`transcriptsBy()`, `exonsBy()`, `cdsBy()`, `intronsByTranscript()`, 
`fiveUTRsByTranscript()`, `threeUTRsByTranscript()`. Two versions of results 
are provided: _tibble_ (`transcripts_tbl()`) and _GRanges_ or _GRangesList_ 
(`transcripts()`). 

Filter can be applied to all extractor functions. A named list of vectors can 
be used to restrict the output, valid names come from any field name from 
tables displayed by `src_tbls()`. Besides, 'granges' could also be used as 
filter for the methods with result displaying as _GRanges_ or _GRangesList_. 

```{r}
filters <- list(SymbolFilter(c("PTEN", "BRCA1")),
                EntrezFilter(5728), 
                GRangesFilter(as("chr10:87869000-87876000", "GRanges")))
transcripts_tbl(src, filter=filters)
transcripts(src, filter=filters)
```

Transcript coordinates of all gene symbols start with "BRCA"

```{r}
transcripts(src, filter = list(SymbolFilter("BRCA", "startsWith")))
```


```{r}
sessionInfo()
```



[dplyr]: https://CRAN.R-project.org/package=dplyr
